name: Merge check for pull requests

on:
  pull_request:
    types: ["opened", "reopened", "edited", "synchronize", "ready_for_review"]

jobs:
  merge-check:
    runs-on: ubuntu-latest

    steps:
      - name: Checkout
        uses: actions/checkout@v3

      - name: Set up Node
        uses: actions/setup-node@v3
        with:
          node-version: 18
          cache: 'npm'

      - name: Install dependencies
        run: npm install

      - name: Run tests
        run: npm test
      
      - name: Review Code Quality
        run: npm run lint

      - name: Run build
        run: npm run build

      - name: The job has failed
        if: ${{ failure() }}
        run: 
          echo "::error::Pull Request cannot be merged due to failed checks." && exit 1
  
  storybook-check:
    timeout-minutes: 60
    runs-on: ubuntu-22.04
    steps:
      - name: Checkout
        uses: actions/checkout@v3

      - name: Set up Node
        uses: actions/setup-node@v3
        with:
          node-version: 18

      - name: Install dependencies
        run: npm install

      - name: Install Playwright
        run: npx playwright install --with-deps

      - name: Build Storybook
        run: npm run build-storybook --quiet

      - name: Serve Storybook and run tests
        run: |
          npx concurrently -k -s first -n "SB,TEST" -c "magenta,blue" \
            "npx http-server storybook-static --port 6006 --silent" \
            "npx wait-on tcp:127.0.0.1:6006 && npm run test-storybook"
      
      - name: The job has failed
        if: ${{ failure() }}
        run: 
          echo "::error::Pull Request cannot be merged due to failed storybook check." && exit 1
